File: java/src/tarehart/rlbot/carpredict/CarSlice.java
package tarehart.rlbot.carpredict;

import tarehart.rlbot.input.CarOrientation;
import tarehart.rlbot.math.SpaceTime;
import tarehart.rlbot.math.vector.Vector3;
import tarehart.rlbot.time.GameTime;

public class CarSlice {
    public final Vector3 space;
    public final GameTime time;
    public final Vector3 velocity;
    public final CarOrientation orientation;

    public CarSlice(Vector3 space, GameTime time, Vector3 velocity, CarOrientation orientation) {
        this.space = space;
        this.time = time;
        this.velocity = velocity;
        this.orientation = orientation;
    }

    public Vector3 getSpace() {
        return space;
    }

    public Vector3 getVelocity() {
        return velocity;
    }

    public GameTime getTime() {
        return time;
    }

    public SpaceTime toSpaceTime() {
        return new SpaceTime(space, time);
    }

    @Override
    public String toString() {
        return "CarSlice{" +
                "space=" + space +
                ", time=" + time +
                ", velocity=" + velocity +
                ", orientation=" + orientation +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        CarSlice carSlice = (CarSlice) o;

        if (space != null ? !space.equals(carSlice.space) : carSlice.space != null) return false;
        if (time != null ? !time.equals(carSlice.time) : carSlice.time != null) return false;
        if (velocity != null ? !velocity.equals(carSlice.velocity) : carSlice.velocity != null) return false;
        return orientation != null ? orientation.equals(carSlice.orientation) : carSlice.orientation == null;
    }

    @Override
    public int hashCode() {
        int result = space != null ? space.hashCode() : 0;
        result = 31 * result + (time != null ? time.hashCode() : 0);
        result = 31 * result + (velocity != null ? velocity.hashCode() : 0);
        result = 31 * result + (orientation != null ? orientation.hashCode() : 0);
        return result;
    }
}

File: src/main/java/tarehart/rlbot/carpredict/CarSlice.kt
package tarehart.rlbot.carpredict

import rlbot.render.Renderer
import tarehart.rlbot.input.CarData
import tarehart.rlbot.input.CarHitbox
import tarehart.rlbot.input.CarOrientation
import tarehart.rlbot.math.SpaceTime
import tarehart.rlbot.math.vector.Vector3
import tarehart.rlbot.time.GameTime
import java.awt.Color

data class CarSlice(
        val space: Vector3,
        val time: GameTime,
        val velocity: Vector3,
        val orientation: CarOrientation,
        val hitbox: CarHitbox = CarHitbox.OCTANE) {

    constructor(carData: CarData): this(carData.position, carData.time, carData.velocity, carData.orientation, carData.hitbox)

    fun toSpaceTime(): SpaceTime {
        return SpaceTime(space, time)
    }

    fun render(renderer: Renderer, color: Color) {
        val toRoof = orientation.roofVector * (hitbox.height / 2)
        val toNose = orientation.noseVector * (hitbox.length / 2)
        val toSide = orientation.rightVector * (hitbox.width / 2)

        val localOffset = orientation.matrix.dot(hitbox.offset)
        val overallOffset = space + localOffset

        renderer.drawLine3d(color, toRoof + toNose + toSide + overallOffset, toRoof + toNose - toSide + overallOffset)
        renderer.drawLine3d(color, toRoof + toNose + toSide + overallOffset, toRoof - toNose + toSide + overallOffset)
        renderer.drawLine3d(color, toRoof + toNose - toSide + overallOffset, toRoof - toNose - toSide + overallOffset)
        renderer.drawLine3d(color, toRoof - toNose + toSide + overallOffset, toRoof - toNose - toSide + overallOffset)
    }
}
